### PR TITLE
Task Updates, Pixel Splitting, Various Fixes

### PR DESCRIPTION
**Pull request by trahxam**
- Pixel splitting experiment
- Gaussian NLL regression task
- Object classification task
- Pooling option in maskformer
- Add attention bias option for edge features
- Support doing input embedding at higher precision
- Various small fixes

### CODE REVIEW

File: src/hepattn/models/matcher.py [hunk 1]
```diff
@@ -137,6 +136,7 @@ def adapt_solver(self, costs):
             start_time = time.time()
             self.compute_matching(costs)
             solver_times[solver_name] = time.time() - start_time
+            print(f"Adaptive LAP Solver: Evaluated {solver_name}, took {solver_times[solver_name]:.2f}s")
```
[the remark this answers is not in the document]
Comment (samvanstroud) [reply src/hepattn/models/matcher.py:139]: Agree here, not sure we want to print this by default

File: pyproject.toml [hunk 2]
```diff
@@ -15,19 +15,19 @@ dependencies = [
     "ruff==0.8.0",
     "pre-commit>=4.0.1,<5",
     "matplotlib==3.9.*",
-    "lightning>=2.4.0,<3",
+    "lightning==2.5.0.post0",
```
Comment (samvanstroud) [inline pyproject.toml]: I think this is kicking the can down the road no? At some point we will want to update, and when we do the ckpt path stuff will break again

File: src/hepattn/models/maskformer.py [hunk 3]
```diff
@@ -115,7 +121,11 @@ def forward(self, inputs: dict[str, Tensor]) -> dict[str, Tensor]:
 
         # Generate the queries that represent objects
         x["query_embed"] = self.query_initial.expand(batch_size, -1, -1)
-        x["query_valid"] = torch.full((batch_size, self.num_queries), True)
+        x["query_valid"] = torch.full((batch_size, self.num_queries), True, device=x["query_embed"].device)
+
+        # Do any pooling if desired
+        if self.pooling is not None:
+            x = x | self.pooling(x)
```
[the remark this answers is not in the document]
Comment (trahxam) [reply src/hepattn/models/maskformer.py:128]: Although one thing - this means we will have to specify the objects which will be pooled over to the mask former config, which is arguably less clean

File: src/hepattn/__init__.py [hunk 4]
```diff
@@ -1,3 +0,0 @@
-import torch
-
-torch.set_float32_matmul_precision("medium")
```
[the remark this answers is not in the document]
Comment (trahxam) [reply src/hepattn/__init__.py:3]: yeah agree

File: src/hepattn/models/maskformer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (samvanstroud) [reply src/hepattn/models/maskformer.py:128]: Yeah this situation comes up often. I think there are a couple of viable options:

- `x['thing_pooled'] = self.pooling(x)` (let the pooling module extract the relevant info)
- `x['thing_pooled'] = self.pooling(x[self.pooling.input_name])` (extract the relevant info when calling)

I don't have a strong preference, the first looks more clean, but the second is more explicit, modular and easier to document and test)

File: src/hepattn/models/matcher.py [hunk 5]
```diff
@@ -6,27 +6,31 @@
...
 
-def solve_1015_early(cost):
-    return lap1015.lap_early(cost)
+SOLVERS = {
+    "scipy": solve_scipy,
+}
 
+# Some compiled extension can cause SIGKILL errors if compiled for the wrong arch
+# So we have to check they won't kill everything when we import them
+if check_import_safe("lap1015"):
```
Comment (samvanstroud) [inline src/hepattn/models/matcher.py:23]: can we raise a warning that `lap1015` will be disabled in the `else` case, and suggest that the user reinstalls the python package on their new arch?

File: src/hepattn/utils/cli.py [hunk 6]
```diff
@@ -92,9 +108,22 @@ def before_instantiate_classes(self) -> None:
...
+            if "ckpt_path" in sc:
+                self.trainer.ckpt_path = sc["ckpt_path"]
+                self.trainer.ckpt_dir = Path(self.trainer.ckpt_path).parent
+                self.trainer.ckpt_name = str(Path(self.trainer.ckpt_path).stem)
+
+            # After instantiating classes, set the checkpoint path if not provided
+            if not self.trainer.ckpt_path:
+                config = self.config[self.subcommand]["config"]
+                assert len(config) == 1
+                self.trainer.ckpt_path = get_best_epoch(Path(config[0].rel_path))
```
Comment (samvanstroud) [inline src/hepattn/utils/cli.py:129]: is there some redundancy here?

File: src/hepattn/utils/import_utils.py [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+def check_import_safe(module_name):
+    code = f"""
+try:
+    import {module_name}
+except Exception:
+    exit(1)
+exit(0)
+"""
+    result = subprocess.run([sys.executable, "-c", code])
```
Comment (samvanstroud) [inline src/hepattn/utils/import_utils.py:13]: do you really need a `subprocess.run` here rather than just `try`

Comment (trahxam) [reply src/hepattn/utils/import_utils.py:13]: I think so because the crash is a low level CPU-level error, not a python exception, so it is not caught by a `try` block.

File: src/hepattn/utils/cli.py (hunk 6 above)
Comment (trahxam) [reply src/hepattn/utils/cli.py:129]: in terms of defining the dir and name as opposed to just the path?

Comment (trahxam) [reply src/hepattn/utils/cli.py:129]: the dir and name is a hack because somewhere lightning sets ckpt_path to be none, which causes the annoying error during testing. the second block assigns the ckpt_path to be the one with best loss if it was not given explicitly?

Comment (samvanstroud) [reply src/hepattn/utils/cli.py:126]: Right, but for me the error was resolved by setting the `trainer.ckpt_path` only, I didn't need to also specify `ckpt_dir` or `ckpt_name`.

Putting it another way, I'm just trying to understand why we don't also need to set dir and name for the case where the ckpt was not specified explicitly. wouldn't something like 

```python
ckpt_path = sc["ckpt_path"] or get_best_epoch(Path(config[0].rel_path))
self.trainer.ckpt_path = ckpt_path
self.trainer.ckpt_dir = Path(self.trainer.ckpt_path).parent
...
```

if the name and dir are really needed?

Comment (samvanstroud) [reply src/hepattn/utils/cli.py:126]: Ah I see, so you were seeing a slightly different issue to me I guess! Thanks.

File: src/hepattn/models/task.py [hunk 8]
```diff
@@ -128,7 +129,7 @@ class HitFilterTask(Task):
     def __init__(
         self,
         name: str,
-        input_object: str,
+        hit_name: str,
```
[the remark this answers is not in the document]
Comment (trahxam) [reply src/hepattn/models/task.py]: Yeah think we should keep it the same for now so everything is consistent. I agree we need a better / more generic name over "hit". "Constituent" is too long. "input" is confusing as some tasks take object/query items as inputs.

File: src/hepattn/models/task.py [hunk 9]
```diff
@@ -404,8 +580,89 @@ def latent(self, x: dict[str, Tensor], pads: dict[str, Tensor] | None = None) ->
...
-        # If padding data is provided, use it to zero out predictions for any hit slots that are not valid
-        if pads is not None:
-            # Shape of padding goes BM -> B1M -> B1M1 -> BNMD
-            x_obj_hit = x_obj_hit * pads[self.input_hit + "_valid"].unsqueeze(-2).unsqueeze(-1).expand_as(x_obj_hit).float()
+        # Shape of padding goes BM -> B1M -> B1M1 -> BNMD
+        x_obj_hit = x_obj_hit * x[self.input_hit + "_valid"].unsqueeze(-2).unsqueeze(-1).expand_as(x_obj_hit).float()
         return x_obj_hit
+
+
+class ClassificationTask(Task):
```
[the remark this answers is not in the document]
Comment (trahxam) [reply src/hepattn/models/task.py:588]: Yeah this was added in mostly so I could do some classification without messing with every other experiment. After this PR I can combine them into one like we had in the older repo?

File: src/hepattn/models/task.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (samvanstroud) [reply src/hepattn/models/task.py]: In the absence of a better suggestion I would go with constituent for now, I think it's fine if things take up a handful more characters if they are more clear. Here's an issue: https://github.com/samvanstroud/hepattn/issues/65
